crates/corro-types/src/updates.rs

```diff
@@ -0,0 +1,560 @@
+                match_count += 1;
+            }
+        }
+
+        // metrics...
+        for (table, pks) in candidates.clone() {
+            handle
+                .get_counter(&table)
+                .matched_count
+                .increment(pks.len() as u64);
+        }
+
+        trace!(sub_id = %id, %db_version, "found {match_count} candidates");
+
+        if let Err(e) = handle
+            .changes_tx()
+            .try_send((candidates.clone(), db_version))
+        {
+            error!(sub_id = %id, "could not send change candidates to {trait_type} handler: {e}");
+            match e {
+                mpsc::error::TrySendError::Full(item) => {
+                    warn!("channel is full, falling back to async send");
+
+                    let changes_tx = handle.changes_tx().clone();
+                    tokio::spawn(async move {
+                        _ = changes_tx.send(item).await;
+                    });
+                }
+                mpsc::error::TrySendError::Closed(_) => {
+                    if let Some(handle) = manager.remove(id) {
+                        tokio::spawn(async move {
+                            handle.cleanup().await;
+                        });
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub fn match_changes_from_db_version<H>(
+    manager: &impl Manager<H>,
+    conn: &Connection,
+    db_version: CrsqlDbVersion,
+) -> rusqlite::Result<()>
+where
+    H: Handle + Send + 'static,
+{
+    let handles = manager.get_handles();
+    if handles.is_empty() {
+        return Ok(());
+    }
+
+    let trait_type = manager.trait_type();
+    let mut candidates = handles
+        .iter()
+        .map(|(id, handle)| (id, (MatchCandidates::new(), handle)))
+        .collect::<BTreeMap<_, _>>();
+
+    {
+        let mut prepped = conn.prepare_cached(
+            r#"
+        SELECT "table", pk, cid, cl
+            FROM crsql_changes
+            WHERE db_version = ?
+            ORDER BY seq ASC
+        "#,
+        )?;
+
+        let rows = prepped.query_map([db_version], |row| {
+            Ok((
+                row.get::<_, TableName>(0)?,
+                row.get::<_, Vec<u8>>(1)?,
+                row.get::<_, ColumnName>(2)?,
+                row.get::<_, i64>(3)?,
+            ))
+        })?;
+
+        for change_res in rows {
+            let (table, pk, column, cl) = change_res?;
+
+            for (_id, (candidates, handle)) in candidates.iter_mut() {
+                let change = MatchableChange {
+                    table: &table,
+                    pk: &pk,
+                    column: &column,
+                    cl,
+                };
+                handle.filter_matchable_change(candidates, change);
+            }
+        }
+    }
+
+    // metrics...
+    for (id, (candidates, handle)) in candidates {
+        let mut match_count = 0;
+        for (table, pks) in candidates.clone() {
```

Comment:
Would `.iter()` be sufficient instead of `.clone()`? It doesn't look like you need a owned value in that loop.

Reply:
ah. this was from when I had the associated type. will take it out.